### DISCUSSION TITLE
Strange behavior of record struct constructor with default parameters values.

### DISCUSSION THREAD

**Discussion by AlexMihalev** [dotnet/roslyn#57766] [Q&A]
I don't know if this is a bug or expected behavior, but this example:
```csharp
using System;
					
public class Program
{
	public record class SampleClass(int X = 10)
	{
		public int Y { get; init; } = 42;
	}

	public record struct SampleStruct(int X = 10)
	{
		public int Y { get; init; } = 42;
	}
	
	public static void Main()
	{
		Console.WriteLine($"SampleClass()  : {new SampleClass()}");
		Console.WriteLine($"SampleStruct() : {new SampleStruct()}");

		Console.WriteLine($"SampleClass(11)  : {new SampleClass(11)}");
		Console.WriteLine($"SampleStruct(11) : {new SampleStruct(11)}");
	}
}
```
prints following output:
```
SampleClass()  : SampleClass { X = 10, Y = 42 }
>>> SampleStruct() : SampleStruct { X = 0, Y = 0 }
SampleClass(11)  : SampleClass { X = 11, Y = 42 }
SampleStruct(11) : SampleStruct { X = 11, Y = 42 }
```
if this is not a bug can someone point me to where this is documented?
I just try to understand where the record struct can be used as a drop-in replacement for records introduced in the previous framework version.

[Here] (https://dotnetfiddle.net/WZceCU) link to .Net Fiddle

**Comment by RikkiGibson** [comment]
It seems like this has always been a bit of a footgun. If you call the constructor without giving arguments for the optional parameters, then the "default" parameterless struct constructor (which is the same thing as the `default` literal) is a better match than the single-parameter constructor.

A similar behavior can be observed below. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChiA3CZAAgGdKBeSmAUwHdKBlBNAYwQAoAlAG5iAYQD2MauIA2TAHQB1ZFARM+1eQA9hxYlgDMNLql4cTvYgG9ilO5UOVYCSlpFF7Do5x79nr+ko8AUorSgQACyhNLUC3SgBfYiSiIA)

```cs
using System;

var s = new Struct();
Console.Write(s.x);

public struct Struct
{
    public int x;
    public Struct(int x = 1) { this.x = x; }
}
```

I don't know if there is any plan to address this in general, i.e. a warning like "this method with optional parameters is only callable when passing an argument for an optional parameter. Did you meant to do that?". But calling the default struct constructor when you meant to call a "real" constructor seems like a particularly painful result.

Do we have any existing issue for this @jmarolf @jcouv?

**Reply by jcouv** [reply]
I'm not aware of an existing issue. This problem has existed with optional parameters all along (example: `void M() { }` `void M(int i = 0) { }`).

**Reply by AlexMihalev** [reply]
But if I understand correctly before C# 10, structs can't have a parameterless constructor which is weird (for me) but at least somehow explains this historical behavior (from the caller side).

In C# 10 parameterless constructors became completely valid (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/parameterless-struct-constructors) and the difference between A and B became less obvious:
```csharp
using System;

var a = new A();
Console.WriteLine(a.x);

var b = new B();
Console.WriteLine(b.x);

public struct A
{
    public int x = 1;
    public A() {}
}

public struct B
{
    public int x;
    public B(int x = 1) { this.x = x; }
}
```

**Reply by jcouv** [reply]
Yes, in C# 10, it is allowed to declare a parameterless struct constructor. But a default one still exists (although not emitted in metadata/IL) even when not explicitly declared.
This falls out of previous rules: "Instead, every struct implicitly has a parameterless instance constructor which always returns the value that results from setting all value type fields to their default value and all reference type fields to null." ([spec](https://github.com/dotnet/csharplang/blob/main/spec/structs.md#default-values))

**Reply by AlexMihalev** [reply]
So now the default(Struct) will call this default parameter constructor, and the new Struct() will call a custom parameterless constructor if it is defined. But what if I have a constructor where all parameters have default values - I can't call it, I am right? (as I understand this is pre C# 10 issue). But now it's also hiding field initializers. Here example:
```csharp
using System;

Console.WriteLine($"{new A().x} vs {new B().x} vs {new B(3).x} vs {new B(3.3).x}");

public struct A
{
    public int x = 1;
    public A() { }
}

public struct B
{
    public int x = 1;
    public B(int x = 2) { this.x = x; }
    public B(double x) { }
}
```

**Reply by CyrusNajmabadi** [reply]
>So now the default(Struct) will call this default parameter constructor

I don't believe this to be the case (but i could be wrong).  I would expect `default(T)` to always give you the default value and not a call to any constructor.  @jcouv ?

**Reply by AlexMihalev** [reply]
Maybe I put it wrong, аll what a trying to say that before C# 10 `default(T)` was always equal to `new T()` in case of struct, but now it depends and in the case when T has a constructor with default parameters and fields initializers it again equals to `default(T)`. Which in my opinion is not very obvious - and should be somewhere documented.

**Reply by RikkiGibson** [reply]
It feels like you might be able to write an analyzer for this by doing something like:

1. Start with some method with optional parameters.
2. Create syntax for a call to it with `default(ParameterType)` values as arguments for all the non-optional parameters.
2. Speculatively bind the call.
3. If the call binds to the same method symbol you started with, then everything is good. If it binds to a different method symbol, give a diagnostic. That would mean you expected `M(int x = 0)` to get called but instead `M()` got called, for example.

I'm curious if anything like this has already been considered.
